core/node/base_token_adjuster/src/base_token_l1_behaviour.rs

```diff
@@ -0,0 +1,326 @@
+use std::{
+    cmp::max,
+    ops::{Div, Mul},
+    sync::{Arc, RwLock},
+    time::Instant,
+};
+
+use anyhow::Context;
+use bigdecimal::{BigDecimal, Zero};
+use zksync_config::BaseTokenAdjusterConfig;
+use zksync_eth_client::{BoundEthInterface, CallFunctionArgs, Options};
+use zksync_node_fee_model::l1_gas_price::TxParamsProvider;
+use zksync_types::{
+    base_token_ratio::BaseTokenAPIRatio,
+    ethabi::{Contract, Token},
+    web3::{contract::Tokenize, BlockNumber},
+    Address, U256,
+};
+
+use crate::metrics::{OperationResult, OperationResultLabels, METRICS};
+
+#[derive(Debug, Clone)]
+pub struct UpdateOnL1Params {
+    pub eth_client: Box<dyn BoundEthInterface>,
+    pub gas_adjuster: Arc<dyn TxParamsProvider>,
+    pub token_multiplier_setter_account_address: Address,
+    pub chain_admin_contract: Contract,
+    pub getters_facet_contract: Contract,
+    pub diamond_proxy_contract_address: Address,
+    pub chain_admin_contract_address: Option<Address>,
+    pub config: BaseTokenAdjusterConfig,
+    pub last_persisted_l1_ratio: Arc<RwLock<Option<BigDecimal>>>,
```

Comment:
1. Arc<RwLock> is an overkill in this case, you can just use `Option<BigDecimal>` and mutate it (`&self` should be changed to `&mut self` in `retry_update_l1`/`update_l1`
2. I think it would be nice to move `last_persisted_l1_ratio` out of `UpdateOnL1Params` and make it a field of `UpdateOnL1` enum variant. This way params will be immutable part, and `last_persisted_l1_ratio` will be changed on each iteration

Reply:
good points - addressed both